File: android/lib/src/main/kotlin/com/noise/protocol/pattern/HandshakePattern.kt
package com.noise.protocol.pattern

/** Token types in a Noise handshake message pattern. */
enum class Token {
    E, S, EE, ES, SE, SS, PSK
}

/** A message pattern is a sequence of tokens processed in one direction. */
typealias MessagePattern = List<Token>

/**
 * Defines a Noise handshake pattern with pre-messages and message patterns.
 */
data class HandshakePattern(
    val name: String,
    val initiatorPreMessage: List<Token> = emptyList(),
    val responderPreMessage: List<Token> = emptyList(),
    val messagePatterns: List<MessagePattern>
) {
    /** Apply PSK modifiers. Position 0 = start of first message, N = end of Nth message. */
    fun withPSK(positions: List<Int>): HandshakePattern {
        val modified = messagePatterns.map { it.toMutableList() }
        for (pos in positions.sorted()) {
            if (pos == 0) {
                modified[0].add(0, Token.PSK)
            } else {
                val msgIndex = pos - 1
                if (msgIndex < modified.size) {
                    modified[msgIndex].add(Token.PSK)
                }
            }
        }
        val pskSuffix = positions.joinToString("+") { "psk$it" }
        return copy(
            name = "$name$pskSuffix",
            messagePatterns = modified
        )
    }

    companion object {
        // One-way patterns
        val N = HandshakePattern("N", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.ES)))
        val K = HandshakePattern("K", initiatorPreMessage = listOf(Token.S),
            responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.ES, Token.SS)))
        val X = HandshakePattern("X", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.ES, Token.S, Token.SS)))

        // Fundamental interactive patterns
        val NN = HandshakePattern("NN", messagePatterns = listOf(
            listOf(Token.E), listOf(Token.E, Token.EE)))
        val NK = HandshakePattern("NK", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.ES), listOf(Token.E, Token.EE)))
        val NX = HandshakePattern("NX", messagePatterns = listOf(
            listOf(Token.E), listOf(Token.E, Token.EE, Token.S, Token.ES)))
        val XN = HandshakePattern("XN", messagePatterns = listOf(
            listOf(Token.E), listOf(Token.E, Token.EE), listOf(Token.S, Token.SE)))
        val XK = HandshakePattern("XK", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.ES), listOf(Token.E, Token.EE),
                listOf(Token.S, Token.SE)))
        val XX = HandshakePattern("XX", messagePatterns = listOf(
            listOf(Token.E), listOf(Token.E, Token.EE, Token.S, Token.ES),
            listOf(Token.S, Token.SE)))
        val KN = HandshakePattern("KN", initiatorPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E), listOf(Token.E, Token.EE, Token.SE)))
        val KK = HandshakePattern("KK", initiatorPreMessage = listOf(Token.S),
            responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.ES, Token.SS),
                listOf(Token.E, Token.EE, Token.SE)))
        val KX = HandshakePattern("KX", initiatorPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E),
                listOf(Token.E, Token.EE, Token.SE, Token.S, Token.ES)))
        val IN = HandshakePattern("IN", messagePatterns = listOf(
            listOf(Token.E, Token.S), listOf(Token.E, Token.EE, Token.SE)))
        val IK = HandshakePattern("IK", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.ES, Token.S, Token.SS),
                listOf(Token.E, Token.EE, Token.SE)))
        val IX = HandshakePattern("IX", messagePatterns = listOf(
            listOf(Token.E, Token.S),
            listOf(Token.E, Token.EE, Token.SE, Token.S, Token.ES)))

        // Deferred patterns
        val NK1 = HandshakePattern("NK1", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E), listOf(Token.E, Token.EE, Token.ES)))
        val NX1 = HandshakePattern("NX1", messagePatterns = listOf(
            listOf(Token.E), listOf(Token.E, Token.EE, Token.S), listOf(Token.ES)))
        val X1N = HandshakePattern("X1N", messagePatterns = listOf(
            listOf(Token.E), listOf(Token.E, Token.EE), listOf(Token.S), listOf(Token.SE)))
        val X1K = HandshakePattern("X1K", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.ES), listOf(Token.E, Token.EE),
                listOf(Token.S), listOf(Token.SE)))
        val XK1 = HandshakePattern("XK1", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E), listOf(Token.E, Token.EE, Token.ES),
                listOf(Token.S, Token.SE)))
        val X1K1 = HandshakePattern("X1K1", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E), listOf(Token.E, Token.EE, Token.ES),
                listOf(Token.S), listOf(Token.SE)))
        val X1X = HandshakePattern("X1X", messagePatterns = listOf(
            listOf(Token.E), listOf(Token.E, Token.EE, Token.S, Token.ES),
            listOf(Token.S), listOf(Token.SE)))
        val XX1 = HandshakePattern("XX1", messagePatterns = listOf(
            listOf(Token.E), listOf(Token.E, Token.EE, Token.S),
            listOf(Token.ES, Token.S, Token.SE)))
        val X1X1 = HandshakePattern("X1X1", messagePatterns = listOf(
            listOf(Token.E), listOf(Token.E, Token.EE, Token.S),
            listOf(Token.ES, Token.S), listOf(Token.SE)))
        val K1N = HandshakePattern("K1N", initiatorPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E), listOf(Token.E, Token.EE), listOf(Token.SE)))
        val K1K = HandshakePattern("K1K", initiatorPreMessage = listOf(Token.S),
            responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.ES), listOf(Token.E, Token.EE),
                listOf(Token.SE)))
        val KK1 = HandshakePattern("KK1", initiatorPreMessage = listOf(Token.S),
            responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E), listOf(Token.E, Token.EE, Token.SE, Token.ES)))
        val K1K1 = HandshakePattern("K1K1", initiatorPreMessage = listOf(Token.S),
            responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E), listOf(Token.E, Token.EE, Token.ES),
                listOf(Token.SE)))
        val K1X = HandshakePattern("K1X", initiatorPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E),
                listOf(Token.E, Token.EE, Token.S, Token.ES), listOf(Token.SE)))
        val KX1 = HandshakePattern("KX1", initiatorPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E),
                listOf(Token.E, Token.EE, Token.SE, Token.S), listOf(Token.ES)))
        val K1X1 = HandshakePattern("K1X1", initiatorPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E), listOf(Token.E, Token.EE, Token.S),
                listOf(Token.SE, Token.ES)))
        val I1N = HandshakePattern("I1N", messagePatterns = listOf(
            listOf(Token.E, Token.S), listOf(Token.E, Token.EE), listOf(Token.SE)))
        val I1K = HandshakePattern("I1K", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.ES, Token.S),
                listOf(Token.E, Token.EE), listOf(Token.SE)))
        val IK1 = HandshakePattern("IK1", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.S),
                listOf(Token.E, Token.EE, Token.SE, Token.ES)))
        val I1K1 = HandshakePattern("I1K1", responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(listOf(Token.E, Token.S),
                listOf(Token.E, Token.EE, Token.ES), listOf(Token.SE)))
        val I1X = HandshakePattern("I1X", messagePatterns = listOf(
            listOf(Token.E, Token.S), listOf(Token.E, Token.EE, Token.S, Token.ES),
            listOf(Token.SE)))
        val IX1 = HandshakePattern("IX1", messagePatterns = listOf(
            listOf(Token.E, Token.S), listOf(Token.E, Token.EE, Token.SE, Token.S),
            listOf(Token.ES)))
        val I1X1 = HandshakePattern("I1X1", messagePatterns = listOf(
            listOf(Token.E, Token.S), listOf(Token.E, Token.EE, Token.S),
            listOf(Token.SE, Token.ES)))

        // Fallback - roles reversed from original XX, es/se swapped
        val XXfallback = HandshakePattern("XXfallback",
            responderPreMessage = listOf(Token.E),
            messagePatterns = listOf(
                listOf(Token.E, Token.EE, Token.S, Token.SE),
                listOf(Token.S, Token.ES)))

        // Named PSK patterns
        val NKpsk0 = HandshakePattern("NKpsk0",
            responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(
                listOf(Token.PSK, Token.E, Token.ES),
                listOf(Token.E, Token.EE)))
        val IKpsk2 = HandshakePattern("IKpsk2",
            responderPreMessage = listOf(Token.S),
            messagePatterns = listOf(
                listOf(Token.E, Token.ES, Token.S, Token.SS),
                listOf(Token.E, Token.EE, Token.SE, Token.PSK)))

        /** All named patterns. */
        val all: Map<String, HandshakePattern> = mapOf(
            "N" to N, "K" to K, "X" to X,
            "NN" to NN, "NK" to NK, "NX" to NX,
            "XN" to XN, "XK" to XK, "XX" to XX,
            "KN" to KN, "KK" to KK, "KX" to KX,
            "IN" to IN, "IK" to IK, "IX" to IX,
            "NK1" to NK1, "NX1" to NX1,
            "X1N" to X1N, "X1K" to X1K, "XK1" to XK1, "X1K1" to X1K1,
            "X1X" to X1X, "XX1" to XX1, "X1X1" to X1X1,
            "K1N" to K1N, "K1K" to K1K, "KK1" to KK1, "K1K1" to K1K1,
            "K1X" to K1X, "KX1" to KX1, "K1X1" to K1X1,
            "I1N" to I1N, "I1K" to I1K, "IK1" to IK1, "I1K1" to I1K1,
            "I1X" to I1X, "IX1" to IX1, "I1X1" to I1X1,
            "XXfallback" to XXfallback,
            "NKpsk0" to NKpsk0,
            "IKpsk2" to IKpsk2
        )

        fun named(name: String): HandshakePattern =
            all[name] ?: throw com.noise.protocol.NoiseException.UnknownPattern(name)
    }
}

File: ios/Sources/NoiseProtocol/Pattern/HandshakePattern.swift
import Foundation

/// Token types in a Noise handshake message pattern.
public enum Token: String, Sendable {
    case e, s, ee, es, se, ss, psk
}

/// A message pattern is a sequence of tokens processed in one direction.
public typealias MessagePattern = [Token]

/// Defines a Noise handshake pattern with pre-messages and message patterns.
public struct HandshakePattern: Sendable {
    public let name: String
    /// Pre-message tokens for the initiator (e.g., `[.s]` for K-type patterns).
    public let initiatorPreMessage: [Token]
    /// Pre-message tokens for the responder (e.g., `[.s]` for NK, XK, IK patterns).
    public let responderPreMessage: [Token]
    /// Sequence of message patterns, alternating initiator→responder.
    public let messagePatterns: [MessagePattern]

    public init(
        name: String,
        initiatorPreMessage: [Token] = [],
        responderPreMessage: [Token] = [],
        messagePatterns: [MessagePattern]
    ) {
        self.name = name
        self.initiatorPreMessage = initiatorPreMessage
        self.responderPreMessage = responderPreMessage
        self.messagePatterns = messagePatterns
    }
}

// MARK: - One-way patterns

extension HandshakePattern {
    public static let N = HandshakePattern(
        name: "N",
        responderPreMessage: [.s],
        messagePatterns: [[.e, .es]]
    )

    public static let K = HandshakePattern(
        name: "K",
        initiatorPreMessage: [.s],
        responderPreMessage: [.s],
        messagePatterns: [[.e, .es, .ss]]
    )

    public static let X = HandshakePattern(
        name: "X",
        responderPreMessage: [.s],
        messagePatterns: [[.e, .es, .s, .ss]]
    )
}

// MARK: - Fundamental interactive patterns

extension HandshakePattern {
    public static let NN = HandshakePattern(
        name: "NN",
        messagePatterns: [
            [.e],
            [.e, .ee]
        ]
    )

    public static let NK = HandshakePattern(
        name: "NK",
        responderPreMessage: [.s],
        messagePatterns: [
            [.e, .es],
            [.e, .ee]
        ]
    )

    public static let NX = HandshakePattern(
        name: "NX",
        messagePatterns: [
            [.e],
            [.e, .ee, .s, .es]
        ]
    )

    public static let XN = HandshakePattern(
        name: "XN",
        messagePatterns: [
            [.e],
            [.e, .ee],
            [.s, .se]
        ]
    )

    public static let XK = HandshakePattern(
        name: "XK",
        responderPreMessage: [.s],
        messagePatterns: [
            [.e, .es],
            [.e, .ee],
            [.s, .se]
        ]
    )

    public static let XX = HandshakePattern(
        name: "XX",
        messagePatterns: [
            [.e],
            [.e, .ee, .s, .es],
            [.s, .se]
        ]
    )

    public static let KN = HandshakePattern(
        name: "KN",
        initiatorPreMessage: [.s],
        messagePatterns: [
            [.e],
            [.e, .ee, .se]
        ]
    )

    public static let KK = HandshakePattern(
        name: "KK",
        initiatorPreMessage: [.s],
        responderPreMessage: [.s],
        messagePatterns: [
            [.e, .es, .ss],
            [.e, .ee, .se]
        ]
    )

    public static let KX = HandshakePattern(
        name: "KX",
        initiatorPreMessage: [.s],
        messagePatterns: [
            [.e],
            [.e, .ee, .se, .s, .es]
        ]
    )

    public static let IN = HandshakePattern(
        name: "IN",
        messagePatterns: [
            [.e, .s],
            [.e, .ee, .se]
        ]
    )

    public static let IK = HandshakePattern(
        name: "IK",
        responderPreMessage: [.s],
        messagePatterns: [
            [.e, .es, .s, .ss],
            [.e, .ee, .se]
        ]
    )

    public static let IX = HandshakePattern(
        name: "IX",
        messagePatterns: [
            [.e, .s],
            [.e, .ee, .se, .s, .es]
        ]
    )
}

// MARK: - Deferred patterns

extension HandshakePattern {
    public static let NK1 = HandshakePattern(
        name: "NK1",
        responderPreMessage: [.s],
        messagePatterns: [
            [.e],
            [.e, .ee, .es]
        ]
    )

    public static let NX1 = HandshakePattern(
        name: "NX1",
        messagePatterns: [
            [.e],
            [.e, .ee, .s],
            [.es]
        ]
    )

    public static let X1N = HandshakePattern(
        name: "X1N",
        messagePatterns: [
            [.e],
            [.e, .ee],
            [.s],
            [.se]
        ]
    )

    public static let X1K = HandshakePattern(
        name: "X1K",
        responderPreMessage: [.s],
        messagePatterns: [
            [.e, .es],
            [.e, .ee],
            [.s],
            [.se]
        ]
    )

    public static let XK1 = HandshakePattern(
        name: "XK1",
        responderPreMessage: [.s],
        messagePatterns: [
            [.e],
            [.e, .ee, .es],
            [.s, .se]
        ]
    )

    public static let X1K1 = HandshakePattern(
        name: "X1K1",
        responderPreMessage: [.s],
        messagePatterns: [
            [.e],
            [.e, .ee, .es],
            [.s],
            [.se]
        ]
    )

    public static let X1X = HandshakePattern(
        name: "X1X",
        messagePatterns: [
            [.e],
            [.e, .ee, .s, .es],
            [.s],
            [.se]
        ]
    )

    public static let XX1 = HandshakePattern(
        name: "XX1",
        messagePatterns: [
            [.e],
            [.e, .ee, .s],
            [.es, .s, .se]
        ]
    )

    public static let X1X1 = HandshakePattern(
        name: "X1X1",
        messagePatterns: [
            [.e],
            [.e, .ee, .s],
            [.es, .s],
            [.se]
        ]
    )

    public static let K1N = HandshakePattern(
        name: "K1N",
        initiatorPreMessage: [.s],
        messagePatterns: [
            [.e],
            [.e, .ee],
            [.se]
        ]
    )

    public static let K1K = HandshakePattern(
        name: "K1K",
        initiatorPreMessage: [.s],
        responderPreMessage: [.s],
        messagePatterns: [
            [.e, .es],
            [.e, .ee],
            [.se]
        ]
    )

    public static let KK1 = HandshakePattern(
        name: "KK1",
        initiatorPreMessage: [.s],
        responderPreMessage: [.s],
        messagePatterns: [
            [.e],
            [.e, .ee, .se, .es]
        ]
    )

    public static let K1K1 = HandshakePattern(
        name: "K1K1",
        initiatorPreMessage: [.s],
        responderPreMessage: [.s],
        messagePatterns: [
            [.e],
            [.e, .ee, .es],
            [.se]
        ]
    )

    public static let K1X = HandshakePattern(
        name: "K1X",
        initiatorPreMessage: [.s],
        messagePatterns: [
            [.e],
            [.e, .ee, .s, .es],
            [.se]
        ]
    )

    public static let KX1 = HandshakePattern(
        name: "KX1",
        initiatorPreMessage: [.s],
        messagePatterns: [
            [.e],
            [.e, .ee, .se, .s],
            [.es]
        ]
    )

    public static let K1X1 = HandshakePattern(
        name: "K1X1",
        initiatorPreMessage: [.s],
        messagePatterns: [
            [.e],
            [.e, .ee, .s],
            [.se, .es]
        ]
    )

    public static let I1N = HandshakePattern(
        name: "I1N",
        messagePatterns: [
            [.e, .s],
            [.e, .ee],
            [.se]
        ]
    )

    public static let I1K = HandshakePattern(
        name: "I1K",
        responderPreMessage: [.s],
        messagePatterns: [
            [.e, .es, .s],
            [.e, .ee],
            [.se]
        ]
    )

    public static let IK1 = HandshakePattern(
        name: "IK1",
        responderPreMessage: [.s],
        messagePatterns: [
            [.e, .s],
            [.e, .ee, .se, .es]
        ]
    )

    public static let I1K1 = HandshakePattern(
        name: "I1K1",
        responderPreMessage: [.s],
        messagePatterns: [
            [.e, .s],
            [.e, .ee, .es],
            [.se]
        ]
    )

    public static let I1X = HandshakePattern(
        name: "I1X",
        messagePatterns: [
            [.e, .s],
            [.e, .ee, .s, .es],
            [.se]
        ]
    )

    public static let IX1 = HandshakePattern(
        name: "IX1",
        messagePatterns: [
            [.e, .s],
            [.e, .ee, .se, .s],
            [.es]
        ]
    )

    public static let I1X1 = HandshakePattern(
        name: "I1X1",
        messagePatterns: [
            [.e, .s],
            [.e, .ee, .s],
            [.se, .es]
        ]
    )
}

// MARK: - Fallback pattern

extension HandshakePattern {
    /// XXfallback: roles are reversed from original XX.
    /// The fallback initiator (original responder) sends first.
    /// The responder's pre-message `e` is the original initiator's ephemeral.
    /// DH tokens es/se are swapped to account for role reversal.
    public static let XXfallback = HandshakePattern(
        name: "XXfallback",
        responderPreMessage: [.e],
        messagePatterns: [
            [.e, .ee, .s, .se],
            [.s, .es]
        ]
    )
}

// MARK: - Named PSK patterns

extension HandshakePattern {
    public static let NKpsk0 = HandshakePattern(
        name: "NKpsk0",
        responderPreMessage: [.s],
        messagePatterns: [
            [.psk, .e, .es],
            [.e, .ee]
        ]
    )

    /// IKpsk2 (WireGuard-style): IK with PSK at end of second handshake message.
    public static let IKpsk2 = HandshakePattern(
        name: "IKpsk2",
        responderPreMessage: [.s],
        messagePatterns: [
            [.e, .es, .s, .ss],
            [.e, .ee, .se, .psk]
        ]
    )
}

// MARK: - PSK modifier

extension HandshakePattern {
    /// Apply PSK modifiers to a base pattern.
    /// `positions` indicates where psk tokens are placed:
    /// - 0: beginning of first message
    /// - N (1..): end of Nth message
    public func withPSK(positions: [Int]) -> HandshakePattern {
        var modified = messagePatterns
        for pos in positions.sorted() {
            if pos == 0 {
                modified[0].insert(.psk, at: 0)
            } else {
                let msgIndex = pos - 1
                if msgIndex < modified.count {
                    modified[msgIndex].append(.psk)
                }
            }
        }
        let pskSuffix = positions.map { "psk\($0)" }.joined(separator: "+")
        return HandshakePattern(
            name: "\(name)\(pskSuffix)",
            initiatorPreMessage: initiatorPreMessage,
            responderPreMessage: responderPreMessage,
            messagePatterns: modified
        )
    }
}

// MARK: - Lookup

extension HandshakePattern {
    /// All named fundamental and deferred patterns.
    public static let all: [String: HandshakePattern] = [
        "N": .N, "K": .K, "X": .X,
        "NN": .NN, "NK": .NK, "NX": .NX,
        "XN": .XN, "XK": .XK, "XX": .XX,
        "KN": .KN, "KK": .KK, "KX": .KX,
        "IN": .IN, "IK": .IK, "IX": .IX,
        "NK1": .NK1, "NX1": .NX1,
        "X1N": .X1N, "X1K": .X1K, "XK1": .XK1, "X1K1": .X1K1,
        "X1X": .X1X, "XX1": .XX1, "X1X1": .X1X1,
        "K1N": .K1N, "K1K": .K1K, "KK1": .KK1, "K1K1": .K1K1,
        "K1X": .K1X, "KX1": .KX1, "K1X1": .K1X1,
        "I1N": .I1N, "I1K": .I1K, "IK1": .IK1, "I1K1": .I1K1,
        "I1X": .I1X, "IX1": .IX1, "I1X1": .I1X1,
        "XXfallback": .XXfallback,
        "NKpsk0": .NKpsk0,
        "IKpsk2": .IKpsk2
    ]

    public static func named(_ name: String) throws -> HandshakePattern {
        guard let pattern = all[name] else {
            throw NoiseError.unknownPattern(name)
        }
        return pattern
    }
}
